File: src/mvc/python/crypto_ranking/crypto_ranking_delegate.py
from PyQt6.QtCore import Qt, QSize
from PyQt6.QtGui import QIcon, QColor, QPen
from PyQt6.QtWidgets import QStyledItemDelegate


class CryptoRankingDelegate(QStyledItemDelegate):
    def __init__(self, parent=None):
        super(CryptoRankingDelegate, self).__init__(parent)

    def paint(self, painter, option, index):
        column = index.column()

        if column == 2:
            value = index.data()
            oldPen = painter.pen()
            if value < 0:
                painter.setPen(QPen(QColor("#8b0000")))
            else:
                painter.setPen(QPen(QColor("#006600")))
            icon = index.data(Qt.ItemDataRole.DecorationRole)
            pixmap = icon.pixmap(option.decorationSize)
            painter.drawPixmap(option.rect.x() + 1, option.rect.y() + option.rect.height() / 2 - pixmap.height() / 2, pixmap)
            painter.drawText(option.rect,
                            Qt.AlignmentFlag.AlignCenter,
                         f"{value:.0f}%")
            painter.setPen(oldPen)
        elif column == 1 or column == 3 or column == 4:
            value = index.data()
            rect = option.rect
            rect.setX(rect.x() + 3)
            painter.drawText(rect,
                            option.displayAlignment,
                           f"${value:.0f}%")
        else:
            super().paint(painter, option, index)
    
    def sizeHint(self, option, index):
        originalSize = QStyledItemDelegate.sizeHint(self, option, index)
        return QSize(originalSize.width(), originalSize.height() * 2)
    

File: src/mvc/cpp/crypto_ranking/crypto_ranking_delegate.cpp
#include "crypto_ranking_delegate.h"
#include <QApplication>
#include <QDialog>
#include <QLineEdit>
#include <QPainter>
#include <QProgressBar>
#include <QPushButton>
#include <QVBoxLayout>
#include <QWidget>

CryptoRankingDelegate::CryptoRankingDelegate(QObject *parent)
    : QStyledItemDelegate(parent) {}

CryptoRankingDelegate::~CryptoRankingDelegate() {}

void CryptoRankingDelegate::paint(QPainter *painter,
                                  const QStyleOptionViewItem &option,
                                  const QModelIndex &index) const {
  auto column = index.column();

  if (column == 2) {
    auto value = index.data().toDouble();
    auto oldPen = painter->pen();
    if (value < 0) {
      painter->setPen("#8b0000");
    } else {
      painter->setPen("#006600");
    }
    auto icon = index.data(Qt::DecorationRole).value<QIcon>();
    auto pixmap = icon.pixmap(option.decorationSize);
    painter->drawPixmap(option.rect.x() + 1,
                        option.rect.y() + option.rect.height() / 2 -
                            pixmap.height() / 2,
                        pixmap);
    painter->drawText(option.rect, Qt::AlignCenter,
                      QString::number(value, 'f', 0) + "%");
    painter->setPen(oldPen);
  } else if (column == 1 || column == 3 || column == 4) {
    auto value = index.data().toDouble();
    auto rect = option.rect;
    rect.setX(rect.x() + 3);
    painter->drawText(rect, option.displayAlignment,
                      "$" + QString::number(value, 'f', 0));
  } else {
    QStyledItemDelegate::paint(painter, option, index);
  }
}

QSize CryptoRankingDelegate::sizeHint(const QStyleOptionViewItem &option,
                                      const QModelIndex &index) const {
  // make it twice as long as the default
  auto originalSize = QStyledItemDelegate::sizeHint(option, index);
  return QSize(originalSize.width(), originalSize.height() * 2);
}
